pkg/kubelet/server/stats/fs_resource_analyzer.go

```diff
@@ -0,0 +1,146 @@
+type fsResourceAnalyzer struct {
+	volumeCache         sync.RWMutex
+	statsProvider       StatsProvider
+	cachedVolumeStats   *cachedPodVolumeStats // Eagerly cache volume stats
+	calcVolumeFrequency time.Duration         // How frequently to calculate disk usage
+}
+
+var _ fsResourceAnalyzerInterface = &fsResourceAnalyzer{}
+
+// cachedPodVolumeStats caches the results of eagerly calculating PodVolumeStats
+type cachedPodVolumeStats struct {
+	volumeStats map[types.UID]*PodVolumeStats
+}
+
+// default calculation interval to use if none is provided
+const defaultVolumeStatsCalcInterval = 30 * time.Second
+
+// NewStats returns a new NewDiskResourceAnalyzer implementation
+func newFsResourceAnalyzer(statsProvider StatsProvider, calcVolumeFrequency time.Duration) *fsResourceAnalyzer {
+	return &fsResourceAnalyzer{
+		statsProvider:       statsProvider,
+		calcVolumeFrequency: calcVolumeFrequency,
+	}
+}
+
+// Start eager background caching of volume stats.
+func (s *fsResourceAnalyzer) Start() {
+	go func() {
+		sleepTime := s.calcVolumeFrequency
+		if sleepTime <= 0 {
+			glog.Warningf("volume stats calc interval must be greater than 0, defaulting to %v.", defaultVolumeStatsCalcInterval)
+			sleepTime = defaultVolumeStatsCalcInterval
+		}
+		for {
+			glog.V(3).Infof("Calculating volume stats and caching result.")
+			startTime := time.Now()
+			s.updateCachedPodVolumeStats()
+			elapsedDuration := time.Now().Sub(startTime)
+			glog.V(3).Infof("Finished calculating volume stats in %v, sleeping for %v.", elapsedDuration, sleepTime)
+			time.Sleep(sleepTime)
+		}
+	}()
+}
+
+// updateCachedPodVolumeStats calculates and caches the PodVolumeStats for every Pod known to the kubelet.
+func (s *fsResourceAnalyzer) updateCachedPodVolumeStats() {
+	// Calculate the new volume stats map
+	pods := s.statsProvider.GetPods()
+	newCache := &cachedPodVolumeStats{make(map[types.UID]*PodVolumeStats, len(pods))}
+	for _, pod := range pods {
+		podUid := pod.GetUID()
+		stats, err := s.getPodVolumeStats(podUid)
+		if err != nil {
+			continue
+		}
+		newCache.volumeStats[podUid] = &stats
+	}
+
+	// Update the map reference
+	s.volumeCache.Lock()
+	defer s.volumeCache.Unlock()
+	s.cachedVolumeStats = newCache
+}
+
+// getPodVolumeStats calculates PodVolumeStats for a given pod and returns the result.
+func (s *fsResourceAnalyzer) getPodVolumeStats(uid types.UID) (PodVolumeStats, error) {
+	// Find all Volumes for the Pod
+	volumes, err := s.statsProvider.ListVolumesForPod(uid)
+	if err != nil {
+		return PodVolumeStats{}, err
+	}
+
+	// Call GetMetrics on each Volume and copy the result to a new VolumeStats.FsStats
+	stats := make([]VolumeStats, 0, len(volumes))
+	for name, v := range volumes {
+		metric, err := v.GetMetrics()
+		if err != nil {
+			// Expected for Volumes that don't support Metrics
+			// TODO: Disambiguate unsupported from errors
+			glog.V(3).Infof("Failed to calculate volume metrics for pod %v volume %s: %v", uid, name, err)
```

Comment:
(optional nit) Consider passing in a `*api.Pod` instead of the `uid`, and log `format.Pod(pod)` instead.